File: BFS/913.Cat-and-Mouse/913.Cat-and-Mouse.cpp
class Solution {
    int color[50][50][3];
    
public:
    int catMouseGame(vector<vector<int>>& graph) 
    {
        int N = graph.size();
                
        queue<vector<int>>q;
        for (int i=0; i<N; i++)
            for (int turn=1; turn<=2; turn++)
            {
                color[0][i][turn] = 1;  // mouse win conditions
                q.push({0,i,turn});
                
                if (i!=0)               
                {
                    color[i][i][turn] = 2;  // cat win conditions
                    q.push({i,i,turn});
                }
            }
        
        while (!q.empty())
        {            
            int m = q.front()[0]; 
            int c = q.front()[1];
            int t = q.front()[2];            
            int status = color[m][c][t];
            q.pop();
                        
            for (auto nextNode: findAllParents(graph,m,c,t))
            {
                int m2 = nextNode[0];
                int c2 = nextNode[1];
                int t2 = nextNode[2];
                
                if (color[m2][c2][t2]!=0) continue; // has been determined
                
                if (t2==status)   // immediate win, (m2,c2,t2)->(m,c,t)
                {
                    color[m2][c2][t2] = status;
                    q.push({m2,c2,t2});                    
                }
                else if (allChildrenWin(graph, m2,c2,t2))    // eventually lose
                {
                    color[m2][c2][t2] = (t2==1)? 2:1;
                    q.push({m2,c2,t2});                
                }                
            }
        }
        return color[1][2][1];
    }
    
    
    vector<vector<int>> findAllParents(vector<vector<int>>& graph, int m, int c, int t)
    {
        vector<vector<int>>neighbours;
        if (t==1)
        {
            for (int c_next: graph[c])
                if (c_next!=0) neighbours.push_back({m,c_next,2});
        }
        else
        {
            for (int m_next: graph[m])
                neighbours.push_back({m_next,c,1});
        }
        return neighbours;
    }
    
    bool allChildrenWin(vector<vector<int>>& graph, int m, int c, int t)
    {
        if (t==1)
        {
            for (int m_next: graph[m])
                if (color[m_next][c][2]!=2)
                    return false;
        }
        else if (t==2)
        {
            for (int c_next: graph[c])
                if (c_next!=0 && color[m][c_next][1]!=1)
                    return false;
        }
        return true;
    }
};

File: BFS/913.Cat-and-Mouse/913.Cat-and-Mouse.py
class Solution:
    def catMouseGame(self, graph):
        """
        :type graph: List[List[int]]
        :rtype: int
        """
        N = len(graph)
        color = [[[0]*3 for _ in range(N)] for _ in range(N)]        
        q = collections.deque()
        for i in range(1,N):            
            for t in range(1,3):                
                color[0][i][t] = 1
                q.append((0,i,t))                                                
                color[i][i][t] = 2                
                q.append((i,i,t))
        
        while (q):
            curStatus = q.popleft()            
            cat,mouse,turn = curStatus
            
            for prevStatus in self.findAllPrevStatus(graph,curStatus):
                
                if color[prevStatus[0]][prevStatus[1]][prevStatus[2]]!=0: continue
                
                # immediate win for prevStatus
                if color[cat][mouse][turn]==3-turn:
                    color[prevStatus[0]][prevStatus[1]][prevStatus[2]] = prevStatus[2]
                    q.append(prevStatus)
                
                # forced to lose for prevStatus
                elif self.allNeighboursWin(color,graph,prevStatus):
                    color[prevStatus[0]][prevStatus[1]][prevStatus[2]] = 3-prevStatus[2]
                    q.append(prevStatus)
            
        return color[1][2][1]
    
    def findAllPrevStatus(self,graph,curStatus):
        ret = []
        mouse,cat,turn = curStatus
        if turn==1:
            for prevCat in graph[cat]:
                if prevCat==0: continue
                ret.append((mouse,prevCat,2))
        else:
            for prevMouse in graph[mouse]:
                ret.append((prevMouse,cat,1))
        return ret
    
    def allNeighboursWin(self,color,graph,status):
        mouse,cat,turn = status
        if turn==1:
            for nextMouse in graph[mouse]:                
                if color[nextMouse][cat][2]!=2:
                    return False
        elif turn==2:            
            for nextCat in graph[cat]:
                if nextCat==0: continue
                if color[mouse][nextCat][1]!=1:
                    return False
        return True
